### PR TITLE
Part one upgrade

### PR DESCRIPTION
**Pull request by acald-creator**
**Notes for Reviewers**

This PR is related to the Meshery UI library and packages upgrade at https://github.com/meshery/meshery/issues/6680

**Goal**: Upgrade the following main packages
1. `Next.js` **12** minimum
2. `React@17` & `React-DOM@17`
3. `MUI` **v5** (`@mui/material`)(will open a separate issue to track this)
4. `Eslint` v7 minimum (due to package conflict from `Next.js` **10** to **12**)

**Blockers**: Packages that are conflicted
1. `react-codemirror2` - relies on `React@16`, inactive for 2 years
    1. Suggestions:
        1. upgrade to  `@uiw/react-codemirror` latest version (will require refactoring as YAML linting capabilities is not part of the library)
        2. pin `@uiw/react-codemirror` at `v3` (will not work due to Next.js 11 and above Global CSS support). See error below. And see error referenced at https://nextjs.org/docs/messages/css-npm. **Note that the previously hackity solution within `_app.js` is not working, plus additional suggestions.**
        3. use `codemirror v6` directly as Codemirror have revised their library from `v5` to `v6` and refactor multiple files that call `CodeMirror`
        4. force `react-codemirror2` as `--legacy-peer-deps` until final decision made from above, although this may require a hackity solution in order to make sure this continues to work because of the SSR and Client issues
        5. not force `react-codemirror2` and suggest a different library such as `Monaco Editor` (There is a React version that supports up to React 18)

```bash
./node_modules/@uiw/react-codemirror/esm/codemirror.css
Global CSS cannot be imported from within node_modules.
Read more: https://nextjs.org/docs/messages/css-npm
Location: node_modules/@uiw/react-codemirror/esm/index.js
```

For now, we are testing `monaco-editor` as a replacement, and so far there are several pages that requires changes.
* Applications (have not started)
* Filters (partially done)
* Designs (partially done) - need to fix fullscreen and loading new monaco-editor first

2. `react-select` can be upgraded to `v5` and just need an import changed to `import Creatable from 'react-select/creatable';`
3. `eslint` will need to be pinned at `v7` due to conflicts in between Next.js `10` to `12`
4. Remove `npm uninstall eslint-loader eslint-config-airbnb eslint-config-google eslint-plugin-react-hooks eslint-plugin-jsx-a11y eslint-plugin-import` as these have already been supported internally when `Next` has built-in support for `eslint`. See here https://nextjs.org/docs/basic-features/eslint
5. Add `npm install --save-dev eslint@7.32.0 @next/eslint-plugin-next` and reconfigure `.eslintrc.js` with plugin
6. Update `package.json` scripts to use `next lint`

**MUI v5** (work in progress)
Deprecated message from MUI: `@material-ui/styles@4.11.5: Material UI v4 doesn't receive active development since September 2021. See the guide https://mui.com/material-ui/migration/migration-v4/ to upgrade to v5.`

May decide to do on a separate PR or on same PR to use `@mui/styles` in between migration to `@mui/material`.

Steps taken to upgrade:
1. `npm install react@17 react-dom@17`
2. `npm install next@11` - error received while installing `next@11`

```bash
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: react-codemirror2@7.2.1
npm ERR! Found: react@17.0.2
npm ERR! node_modules/react
npm ERR!   peer react@">=16.3" from @fortawesome/react-fontawesome@0.2.0
npm ERR!   node_modules/@fortawesome/react-fontawesome
npm ERR!     @fortawesome/react-fontawesome@"^0.2.0" from the root project
npm ERR!   peer react@"^16.8.0 || ^17.0.0" from @material-ui/core@4.12.4
npm ERR!   node_modules/@material-ui/core
npm ERR!     peer @material-ui/core@"^4.0.0" from @material-ui/icons@4.11.3
npm ERR!     node_modules/@material-ui/icons
npm ERR!       peer @material-ui/icons@"^4.11.2" from @rjsf/material-ui@5.0.0-beta.11
npm ERR!       node_modules/@rjsf/material-ui
npm ERR!         @rjsf/material-ui@"^5.0.0-beta.11" from the root project
npm ERR!       2 more (mui-datatables, the root project)
npm ERR!     peer @material-ui/core@"^4.12.1" from @material-ui/lab@4.0.0-alpha.61
npm ERR!     node_modules/@material-ui/lab
npm ERR!       @material-ui/lab@"^4.0.0-alpha.61" from the root project
npm ERR!     5 more (@material-ui/pickers, @rjsf/material-ui, ...)
npm ERR!   38 more (@material-ui/icons, @material-ui/lab, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer react@">=15.5 <=16.x" from react-codemirror2@7.2.1
npm ERR! node_modules/react-codemirror2
npm ERR!   react-codemirror2@"^7.2.1" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: react@16.14.0
npm ERR! node_modules/react
npm ERR!   peer react@">=15.5 <=16.x" from react-codemirror2@7.2.1
npm ERR!   node_modules/react-codemirror2
npm ERR!     react-codemirror2@"^7.2.1" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! 
npm ERR! For a full report see:
npm ERR! /home/phxvlabs-io/.npm/_logs/2022-12-19T17_24_17_182Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/phxvlabs-io/.npm/_logs/2022-12-19T17_24_17_182Z-debug-0.log
````
3. `npm uninstall react-countdown-clock react-select react-codemirror2`
4. `npm install next@12`
5. Pin `react-countdown-clock@2.9.0`
6. Pinning `uiw/react-codemirror@v3` & `codemirror@v5` (failing due to Next.js global css errors on third party packages, hackitity solution not working)
7. `npm config set legacy-peer-deps=true --location=project` **This is a temporary solution until we figure out `codemirror and react-codemirror2`**
8. Chose not to rename `<img` to `<Image` due to `Next.js` not optimizing the images properly (**DO NOT DO THIS**, images looked out of wack, will need a solution on how to handle images when using `next export`. Because without changing this, now some of the icons looks huge.)

Future Plans: (Work in Progress)
1. Implement `prettier`, `lint-staged` as `Next.js` has better support
2. Implement `relay` with a configuration file, as `Next.js` has built-in support via `next.config.js`
3. Define a central theme (light/dark) and create reusable components (MUI v5 migration)

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: ui/.npmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+legacy-peer-deps=true
```
Comment (Abhishek-kumar09) [inline ui/.npmrc:1]: is this default behaviour helpful for us?

Comment (Abhishek-kumar09) [reply ui/.npmrc:1]: I see it with fearful eyes.

Comment (acald-creator) [reply ui/.npmrc:1]: I understand, this is just only for this PR, as I would like to resolve `react-codemirror2` first and the fact that the codebase will require a refactor based on this.

I only include this temporarily so I can still see the dashboard locally. Once we resolve this, then I will be able to remove this, when I move onto MUI

**Comment by leecalcote** [comment]
The level of detail and diligence here is confidence-inspiring. @asubedy @Krishna2323 in reference to the challenge with html labels plugin and the suggested paths to resolution, please take note of @acald-creator's attention to and documentation of details.

**Comment by leecalcote** [comment]
Down with `react-codemirror`. What other packages are candidates for its displacement?

**Comment by leecalcote** [comment]
I encourage @acald-creator to signal her championship of the ""[?] Support for Material UI v5, React 17, Nextjs 13
 effort in v0.7.0 here - https://discuss.layer5.io/t/meshery-v0-7-0-roadmap/232

File: ui/package.json [hunk 2]
```diff
@@ -55,7 +54,9 @@
     "@redux-devtools/extension": "^3.2.3",
     "@rjsf/core": "^5.0.0-beta.11",
     "@rjsf/material-ui": "^5.0.0-beta.11",
+    "@rjsf/utils": "^5.0.0-beta.14",
```
Comment (harkiratsm) [inline ui/package.json:57]: let's stick to ^5.0.0-beta.11. [#ref](https://github.com/meshery/meshery/pull/6606#issuecomment-1338729367)

Comment (acald-creator) [reply ui/package.json:57]: @harkiratsm Wasn't intentional, just noticed that the package wasn't installed, I just ran the command to install the package itself, but I can pin this at ^5.0.0-beta.11.

Comment (harkiratsm) [reply ui/package.json:63]: Thanks @acald-creator

**Comment by Abhishek-kumar09** [comment]
When you are pinning a certain package with its version number, make sure to not add "^", it will grab the latest from the list of subversion released of their major version. 

You're already making the list of pinned version numbers and that's great, it will come so handy when we get back.

**Pull request closed without merging**